File: c/primitives.c
#include <inttypes.h>
#include <stdio.h>
#include <string.h>
#include "topple.h"


static Primitive primitives[];

Primitive *find_primitive(const char *name)
{
    for (size_t i = 0; primitives[i].name != NULL; i++) {
        if (strcmp(name, primitives[i].name) == 0)
            return &primitives[i];
    }
    return NULL;
}


#define BINARY(NAME, EXPR)                     \
    static void NAME(Stack *s) {               \
        uint64_t r = num_of_val(stack_pop(s)); \
        uint64_t l = num_of_val(stack_pop(s)); \
        stack_push(s, (EXPR));                 \
    }
#define ARITH(NAME, EXPR) BINARY(NAME, val_of_num(EXPR))
#define COMP(NAME, EXPR) BINARY(NAME, val_of_bool(EXPR))

ARITH(add_, l + r)
ARITH(sub_, l - r)
ARITH(mul_, l * r)
// ARITH(div_, l / r)   -- need to check that (r != 0)
ARITH(shl_, l << (r % 64))
ARITH(shr_, l >> (r % 64))

COMP(eq_, l == r)
COMP(ne_, l != r)
COMP(lt_, l <  r)
COMP(gt_, l >  r)
COMP(le_, l <= r)
COMP(ge_, l >= r)

static void div_(Stack *s)
{
    uint64_t r = num_of_val(stack_pop(s));
    uint64_t l = num_of_val(stack_pop(s));
    if (r == 0)
        fail("division by zero");
    stack_push(s, val_of_num(l / r));
}

#undef COMP
#undef ARITH
#undef BINARY


static void dup_(Stack *s)
{
    Value v = stack_pop(s);
    dup(v);
    stack_push(s, v);
    stack_push(s, v);
}


static void drop_(Stack *s)
{
    discard(stack_pop(s));
}


static void swap_(Stack *s)
{
    Value v1 = stack_pop(s);
    Value v0 = stack_pop(s);
    stack_push(s, v1);
    stack_push(s, v0);
}


static void nip_(Stack *s)
{
    Value v = stack_pop(s);
    drop_(s);
    stack_push(s, v);
}


static void tuck_(Stack *s)
{
    Value v1 = stack_pop(s); dup(v1);
    Value v0 = stack_pop(s);
    stack_push(s, v1);
    stack_push(s, v0);
    stack_push(s, v1);
}


static void over_(Stack *s)
{
    Value v1 = stack_pop(s);
    Value v0 = stack_pop(s); dup(v0);
    stack_push(s, v0);
    stack_push(s, v1);
    stack_push(s, v0);
}


static void rot_(Stack *s)
{
    Value v2 = stack_pop(s);
    Value v1 = stack_pop(s);
    Value v0 = stack_pop(s);
    stack_push(s, v1);
    stack_push(s, v2);
    stack_push(s, v0);
}


static void unrot_(Stack *s)
{
    Value v2 = stack_pop(s);
    Value v1 = stack_pop(s);
    Value v0 = stack_pop(s);
    stack_push(s, v2);
    stack_push(s, v0);
    stack_push(s, v1);
}


static void pick_(Stack *s)
{
    uint64_t n = num_of_val(stack_pop(s));
    Value v = stack_pick(s, n);
    stack_push(s, v);
}


static void dot_(Stack *s)
{
    printf("%"PRIu64" ", num_of_val(stack_pop(s)));
}


static void putc_(Stack *s)
{
    uint64_t c = num_of_val(stack_pop(s));

    if (c < ' ' || c > '~')
        if (c != '\n')
            fail("illegal character");

    printf("%c", (int) c);
}


static void bytes_new_(Stack *s)
{
    Bytes *b = malloc(sizeof(Bytes));
    b->ref_count = 1;
    b->len = 0;
    b->cap = 16;
    b->contents = malloc(16);

    Value v = { .type = BYTES, .bytes = b };
    stack_push(s, v);
}


static void bytes_clear_(Stack *s)
{
    Value b = expect_bytes(stack_pop(s));
    b.bytes->len = 0;
    discard(b);
}


static void bytes_length_(Stack *s)
{
    Value b = expect_bytes(stack_pop(s));
    stack_push(s, val_of_num(b.bytes->len));
    discard(b);
}


static void bytes_push_(Stack *s)
{
    Value b = expect_bytes(stack_pop(s));
    uint64_t c = num_of_val(stack_pop(s));

    if (b.bytes->len == b.bytes->cap) {
        b.bytes->cap *= 2;
        b.bytes = realloc(b.bytes, b.bytes->cap);
    }

    b.bytes->contents[b.bytes->len++] = c % 256;
    discard(b);
}


static void bytes_get_(Stack *s)
{
    Value b = expect_bytes(stack_pop(s));
    uint64_t idx = num_of_val(stack_pop(s));

    if (idx >= b.bytes->len)
        fail("bytes index out of bounds");
    stack_push(s, val_of_num(b.bytes->contents[idx]));
    discard(b);
}


static void bytes_set_(Stack *s)
{
    Value b = expect_bytes(stack_pop(s));
    uint64_t idx = num_of_val(stack_pop(s));
    uint64_t c = num_of_val(stack_pop(s));

    if (idx >= b.bytes->len)
        fail("bytes index out of bounds");
    b.bytes->contents[idx] = c % 256;
    discard(b);
}


static void file_read_(Stack *s)
{
    Value b = expect_bytes(stack_pop(s));
    char *filename = malloc(b.bytes->len + 1);
    memcpy(filename, b.bytes->contents, b.bytes->len);
    filename[b.bytes->len] = '\0';

    FILE *f = fopen(filename, "rb");
    free(filename);
    discard(b);

    if (f == NULL) {
        stack_push(s, val_of_num(0));
        return;
    }

    fseek(f, 0, SEEK_END);
    long size = ftell(f);
    fseek(f, 0, SEEK_SET);

    if (size < 0)
        fail("io error");

    b.type = BYTES;
    b.bytes = malloc(sizeof(Bytes));
    b.bytes->ref_count = 1;
    b.bytes->len = 0;
    b.bytes->cap = size;
    b.bytes->contents = malloc(size);

    unsigned char *buf = b.bytes->contents;
    size_t remaining = size;
    size_t just_read;
    while ((just_read = fread(buf, 1, remaining, f)) > 0) {
        buf += just_read;
        remaining -= just_read;
    }
    b.bytes->len = size - remaining;

    stack_push(s, b);
}


static void block_new_(Stack *s)
{
    Block *b = malloc(sizeof(Block));
    b->ref_count = 1;

    for (uint16_t i = 0; i < 400; i++)
        b->cells[i].type = UNDEFINED;

    Value v;
    v.type = POINTER;
    v.pointer.block = b;
    v.pointer.cell = 0;

    stack_push(s, v);
}


static void pointer_get_(Stack *s)
{
    Value p = expect_pointer(stack_pop(s));
    stack_push(s, var_get(&p.pointer.block->cells[p.pointer.cell]));
    discard(p);
}


static void pointer_set_(Stack *s)
{
    Value p = expect_pointer(stack_pop(s));
    Value v = stack_pop(s);
    var_set(&p.pointer.block->cells[p.pointer.cell], v);
    discard(p);
}


static void pointer_offset_(Stack *s)
{
    uint64_t offset = num_of_val(stack_pop(s));
    Value p = expect_pointer(stack_pop(s));

    offset += (uint64_t) p.pointer.cell;

    if (offset >= 400)
        fail("pointer offset out of bounds");

    p.pointer.cell = offset;
    stack_push(s, p);
}


static Primitive primitives[] = {
    { .name = "+",    .action = add_   },
    { .name = "-",    .action = sub_   },
    { .name = "*",    .action = mul_   },
    { .name = "/",    .action = div_   },
    { .name = "<<",   .action = shl_   },
    { .name = ">>",   .action = shr_   },

    { .name = "=",    .action = eq_    },
    { .name = "<>",   .action = ne_    },
    { .name = "<",    .action = lt_    },
    { .name = ">",    .action = gt_    },
    { .name = "<=",   .action = le_    },
    { .name = ">=",   .action = ge_    },

    { .name = "dup",  .action = dup_   },
    { .name = "drop", .action = drop_  },
    { .name = "swap", .action = swap_  },
    { .name = "nip",  .action = nip_   },
    { .name = "tuck", .action = tuck_  },
    { .name = "over", .action = over_  },
    { .name = "rot",  .action = rot_   },
    { .name = "-rot", .action = unrot_ },
    { .name = "pick", .action = pick_  },

    { .name = ".",    .action = dot_   },
    { .name = "putc", .action = putc_  },

    { .name = "bytes.new",    .action = bytes_new_      },
    { .name = "bytes.clear",  .action = bytes_clear_    },
    { .name = "bytes.length", .action = bytes_length_   },
    { .name = "b%",           .action = bytes_push_     },
    { .name = "b@",           .action = bytes_get_      },
    { .name = "b!",           .action = bytes_set_      },

    { .name = "file.read",    .action = file_read_      },

    { .name = "block.new",    .action = block_new_      },
    { .name = "@",            .action = pointer_get_    },
    { .name = "!",            .action = pointer_set_    },
    { .name = "+p",           .action = pointer_offset_ },

    { .name = NULL },
};

File: src/python/primitives.py
from sys import exit, stderr
from typing import Any

from runtime import ArithPrim, Pointer, Primitive
from util import ToppleException


add = lambda l, r: l + r
sub = lambda l, r: l - r
mul = lambda l, r: l * r
div = lambda l, r: l // r

shl = lambda l, r: l << (r % 64)
shr = lambda l, r: l >> (r % 64)
and_ = lambda l, r: l & r
or_ = lambda l, r: l | r
xor = lambda l, r: l ^ r

eq = lambda l, r: l == r
ne = lambda l, r: l != r
lt = lambda l, r: l < r
gt = lambda l, r: l > r
le = lambda l, r: l <= r
ge = lambda l, r: l >= r


def not_(s):
    with s.pop([int]) as [v]:
        s.append([~v & 0xFFFF_FFFF_FFFF_FFFF])


def dup(s):
    with s.pop([Any]) as [v]:
        s.append([v, v])


def drop(s):
    with s.pop([Any]):
        pass


def swap(s):
    with s.pop([Any, Any]) as [v0, v1]:
        s.append([v1, v0])


def nip(s):
    with s.pop([Any, Any]) as [_, v]:
        s.append([v])


def tuck(s):
    with s.pop([Any, Any]) as [v0, v1]:
        s.append([v1, v0, v1])


def over(s):
    with s.pop([Any, Any]) as [v0, v1]:
        s.append([v0, v1, v0])


def rot(s):
    with s.pop([Any, Any, Any]) as [v0, v1, v2]:
        s.append([v1, v2, v0])


def unrot(s):
    with s.pop([Any, Any, Any]) as [v0, v1, v2]:
        s.append([v2, v0, v1])


def pick(s):
    with s.pop([int]) as [n]:
        v = s.pick(n)
        s.append([v])


def dot(s):
    with s.pop([int]) as [n]:
        print(n, end=" ", file=stderr)


def putc(s):
    with s.pop([int]) as [n]:
        c = chr(n)
        if (c.isascii() and c.isprintable()) or c == " " or c == "\n":
            print(c, end="", file=stderr)
        else:
            raise ToppleException("unprintable character")


def fail(s):
    with s.pop([int]) as [n]:
        exit(n)


def bytes_new(s):
    s.append([bytearray()])


def bytes_clear(s):
    with s.pop([bytearray]) as [b]:
        b.clear()


def bytes_length(s):
    with s.pop([bytearray]) as [b]:
        l = len(b) & 0xFFFF_FFFF_FFFF_FFFF
        s.append([l])


def bytes_push(s):
    with s.pop([int, bytearray]) as [c, b]:
        b.append(c & 0xFF)


def bytes_get(s):
    with s.pop([int, bytearray]) as [idx, b]:
        c = b[idx]
        s.append([c])


def bytes_set(s):
    with s.pop([int, int, bytearray]) as [c, idx, b]:
        b[idx] = c


def file_read(s):
    with s.pop([bytearray]) as [filename]:
        try:
            with open(bytes(filename), "rb") as file:
                contents = file.read()
                s.append([bytearray(contents)])
        except IOError:
            s.append([0])


def block_new(s):
    s.append([Pointer()])


def pointer_get(s):
    with s.pop([Pointer]) as [p]:
        s.append([p.get()])


def pointer_set(s):
    with s.pop([Any, Pointer]) as [v, p]:
        p.set(v)


def pointer_offset(s):
    with s.pop([Pointer, int]) as [p, offset]:
        s.append([p.with_offset(offset)])


primitives = {
    "+": ArithPrim(None, add),
    "-": ArithPrim(None, sub),
    "*": ArithPrim(None, mul),
    "/": ArithPrim(None, div),
    "<<": ArithPrim(None, shl),
    ">>": ArithPrim(None, shr),
    "not": Primitive(None, not_),
    "and": ArithPrim(None, and_),
    "or": ArithPrim(None, or_),
    "xor": ArithPrim(None, xor),
    "=": ArithPrim(None, eq),
    "<>": ArithPrim(None, ne),
    "<": ArithPrim(None, lt),
    ">": ArithPrim(None, gt),
    "<=": ArithPrim(None, le),
    ">=": ArithPrim(None, ge),
    "dup": Primitive(None, dup),
    "drop": Primitive(None, drop),
    "swap": Primitive(None, swap),
    "nip": Primitive(None, nip),
    "tuck": Primitive(None, tuck),
    "over": Primitive(None, over),
    "rot": Primitive(None, rot),
    "-rot": Primitive(None, unrot),
    "pick": Primitive(None, pick),
    ".": Primitive(None, dot),
    "putc": Primitive(None, putc),
    "fail": Primitive(None, fail),
    "bytes.new": Primitive(None, bytes_new),
    "bytes.clear": Primitive(None, bytes_clear),
    "bytes.length": Primitive(None, bytes_length),
    "b%": Primitive(None, bytes_push),
    "b@": Primitive(None, bytes_get),
    "b!": Primitive(None, bytes_set),
    "file.read": Primitive(None, file_read),
    "block.new": Primitive(None, block_new),
    "@": Primitive(None, pointer_get),
    "!": Primitive(None, pointer_set),
    "+p": Primitive(None, pointer_offset),
}
